### PR TITLE
Fixes trait item method DCA warning.

### LINKED ISSUE #3674: `neq` method not used warning when implement `Eq` for a custom type
When implementing `Eq` for a custom type such as:
```rust
enum MyEnum {
    A: u64,
    B: bool,
}

impl core::ops::Eq for MyEnum {
    fn eq(self, other: MyEnum) -> bool {
        match (self, other) {
            (MyEnum::A(inner1), MyEnum::A(inner2))  => inner1 == inner2,
            (MyEnum::B(inner1), MyEnum::B(inner2))  => inner1 == inner2,
            _ => false,
        }
    }
}
```
and only using equality in the program, we still get the following warning:
```console
          warning
             --> /Users/mohammad/Desktop/fuel/sway/sway-lib-core/src/ops.sw:170:5
              |
          168 |
          169 |   } {
          170 |       fn neq(self, other: Self) -> bool {
              |  _____-
          171 | |         (self.eq(other)).not()
          172 | |     }
              | |_____- This method is never called.
          173 |   }
              |
          ____

            Compiled script "configurable_consts" with 1 warning.
```
Even though the warning is true, I don't think we should emit it.

### PR DESCRIPTION
**Pull request by esdrubal**
## Description

```rust
pub trait MyTrait {
    fn trait_method(self) -> bool;
} {
    fn item_method(self) -> bool {
        true
    }
}
```

When a trait implements an item method in a library, the compiler should not throw a warning when the method is not used as it may be used by any Sway program that imports the library.

Closes #3674.

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Review by mohammadfawaz — COMMENTED** [review]
Generally quite reasonable! Thanks. Just some comments.

File: test/src/e2e_vm_tests/test_programs/should_pass/dca/trait_method_lib/json_abi_oracle.json [hunk 1]
```diff
@@ -0,0 +1,25 @@
+{
```
[the remark this answers is not in the document]
Comment (esdrubal) [reply test/src/e2e_vm_tests/test_programs/should_pass/dca/trait_method_lib/json_abi_oracle.json]: Thanks for noticing, removed the 3 `json_abi_oracle.json`.

File: test/src/e2e_vm_tests/test_programs/should_pass/dca/trait_method_lib/src/main.sw [hunk 2]
```diff
@@ -0,0 +1,13 @@
+library trait_method_lib;
+
+pub struct MyStruct {
+
+}
+
+pub trait MyTrait {
+    fn trait_method(self) -> bool;
+} {
+    fn method(self) -> MyStruct {
```
[the remark this answers is not in the document]
Comment (esdrubal) [reply test/src/e2e_vm_tests/test_programs/should_pass/dca/trait_method_lib/src/main.sw:10]: Added example to test case where trait is defined in script file and a warning is thrown.

File: sway-core/src/control_flow_analysis/dead_code_analysis.rs [hunk 3]
```diff
@@ -492,9 +498,21 @@ fn connect_impl_trait<'eng: 'cfg, 'cfg>(
         }
         Some(trait_decl_node) => {
             graph.add_edge_from_entry(entry_node, "".into());
-            graph.add_edge(entry_node, trait_decl_node, "".into());
+            graph.add_edge(entry_node, trait_decl_node.trait_idx, "".into());
         }
     };
+    let trait_entry = graph.namespace.find_trait(trait_name).cloned();
+    let mut trait_items_method_names = Vec::new();
```
[the remark this answers is not in the document]
Comment (esdrubal) [reply sway-core/src/control_flow_analysis/dead_code_analysis.rs:506]: Added comment.

File: sway-core/src/control_flow_analysis/dead_code_analysis.rs [hunk 4]
```diff
@@ -507,7 +525,21 @@ fn connect_impl_trait<'eng: 'cfg, 'cfg>(
...
+                let add_edge_to_fn_decl =
+                    if trait_items_method_names.contains(&fn_decl.name.as_str().to_string()) {
+                        if let Some(trait_entry) = trait_entry.clone() {
+                            matches!(
+                                trait_entry.module_tree_type,
+                                TreeType::Library { .. } | TreeType::Contract
+                            )
+                        } else {
+                            // trait_entry not found which means it probably is compiled in another library
+                            true
```
[the remark this answers is not in the document]
Comment (esdrubal) [reply sway-core/src/control_flow_analysis/dead_code_analysis.rs:540]: Unfortunately when the trait entry is defined in another module we don't have the tree_type so I don't think we can simply this code path. Added better comment there.

**Review by mohammadfawaz — DISMISSED** [review]

**Review by mohammadfawaz — APPROVED** [review]

**Pull request merged**